backends/p4tools/modules/testgen/targets/bmv2/bmv2.cpp

```diff
@@ -43,7 +64,48 @@ CompilerResultOrError Bmv2V1ModelCompilerTarget::runCompilerImpl(
         return std::nullopt;
     }
 
-    return {*new BMv2V1ModelCompilerResult{*program, p4runtimeApi}};
+    // Create DCG.
+    NodesCallGraph *dcg = nullptr;
+    if (TestgenOptions::get().dcg || !TestgenOptions::get().pattern.empty()) {
+        dcg = new NodesCallGraph("NodesCallGraph");
+        P4ProgramDCGCreator dcgCreator(dcg);
+        program->apply(dcgCreator);
+    }
+    if (::errorCount() > 0) {
+        return std::nullopt;
+    }
+    /// Collect coverage information about the program.
+    auto coverage = P4::Coverage::CollectNodes(TestgenOptions::get().coverageOptions);
+    program->apply(coverage);
+    if (::errorCount() > 0) {
+        return std::nullopt;
+    }
```

Comment:
The tricky bit is the ordering because we need to get the P4RuntimeApi before applying the mid end. Since there is only one such example yet it did not seem worthwhile to factor things out into an extra function.

Reply:
Oh, I did not realize that. It would still be better if the DCG & coverage collection did not repeat, but I can live with this for now.